In [63]:
import pickle
import pandas as pd
import os
import glob
from pathlib import Path
import numpy as np

We open the first file of the subset, to check what the HDF5 keys are and then we read each of them.

In [7]:
with pd.HDFStore("data/MillionSongSubset/data/A/A/A/TRAAAAW128F429D538.h5") as hdf:
    print(hdf.keys())

['/analysis/songs', '/metadata/songs', '/musicbrainz/songs']


In [40]:
pd.read_hdf("data/MillionSongSubset/data/A/A/A/TRAAAAW128F429D538.h5","/analysis/songs")

analysis_sample_rate                         audio_md5  danceability  \
0                 22050  a222795e07cd65b7a530f1346f520649           0.0   

    duration  end_of_fade_in  energy  idx_bars_confidence  idx_bars_start  \
0  218.93179           0.247     0.0                    0               0   

   idx_beats_confidence  idx_beats_start         ...          key  \
0                     0                0         ...            1   

   key_confidence  loudness  mode  mode_confidence  start_of_fade_out   tempo  \
0           0.736   -11.197     0            0.636            218.932  92.198   

   time_signature  time_signature_confidence            track_id  
0               4                      0.778  TRAAAAW128F429D538  

[1 rows x 31 columns]

In [21]:
pd.read_hdf("data/MillionSongSubset/data/A/A/A/TRAAAAW128F429D538.h5","/metadata/songs")

analyzer_version  artist_7digitalid  artist_familiarity  artist_hotttnesss  \
0                              165270            0.581794           0.401998   

            artist_id  artist_latitude  artist_location  artist_longitude  \
0  ARD7TVE1187B99BFB1              NaN  California - LA               NaN   

                            artist_mbid artist_name  artist_playmeid genre  \
0  e77e51a5-4761-45b3-9847-2051f811e366      Casual             4479         

   idx_artist_terms  idx_similar_artists      release  release_7digitalid  \
0                 0                    0  Fear Itself              300848   

   song_hotttnesss             song_id             title  track_7digitalid  
0          0.60212  SOMZWCG12A8C13C480  I Didn't Mean To           3401791

In [22]:
pd.read_hdf("data/MillionSongSubset/data/A/A/A/TRAAAAW128F429D538.h5","/musicbrainz/songs")

idx_artist_mbtags  year
0                  0     0

We only need to extract <tt>tempo</tt> and <tt>song_hotttnesss</tt>, here is an example of how to do that on the subset :

In [66]:
tempo = []
hotness = []

files = glob.glob("data/MillionSongSubset/data/A" + "/[A-Z]/[A-Z]/*")
for f in files:
    tempo.append(pd.read_hdf(f,"/analysis/songs")["tempo"][0])
    hotness.append(pd.read_hdf(f,"/metadata/songs")["song_hotttnesss"][0])

In [67]:
tempo = np.asarray(tempo)
hotness = np.asarray(hotness)
print(tempo)
print(hotness)

[ 124.059   80.084   54.874 ...,   96.064  149.012  162.653]
[ 0.54795294  0.47563847         nan ...,         nan         nan  0.270776  ]


In [71]:
print("Number of tracks =", len(files))
print("with missing hotness values =", np.sum(np.isnan(hotness)))

Number of files = 7620
with missing hotness values = 3268


We already got 3268 unknown hotness values and we only tested on a subset of 7620 song, so we can expect to have that information for only a little over half of our final dataset. Maybe we won't use it.

In [64]:
#loading of the dataframe YOU CAN REMOVE THAT IN THE FINAL NOTEBOOK
covers = pickle.load(open("data/covers.p","rb"))
covers

trackID            artistID  shsPerf  clique_id  \
0      TRGDMZP128F42BC52B  ARB1DDF1187FB4FCFB       -1          0   
1      TRCATYW12903D038FE  ARGJEEO1271F573FD6       -1          0   
2      TRVMZJZ128F4270CE4  ARY0HTV1187FB4A1B1       -1          0   
3      TRKOINL128F42926C3  ARQ5FSZ1187B98AD74       -1          0   
4      TROJZTF128F428B546  ARJN76O1187FB43C99       -1          1   
5      TRYQEDQ128F427917C  ARS4KT21187B9B9438       -1          1   
6      TRCKNGE128F92DA3F3  AR1CB5G1187B9AFB8E       -1          2   
7      TRIOPLY128F423CFF3  ARKZJ301187FB521B2       -1          2   
8      TRWNDEU128F9329BF7  ARVZWQ31187B9B8946       -1          3   
9      TRYOPHS128F146DEFD  AR6NYHH1187B9BA128       -1          3   
10     TRMBSQR128F92DF66E  ARPI2DX1187FB4CED4       -1          4   
11     TRQNZCE128E078A9C0  ARWILYB1187FB37DFE       -1          4   
12     TRTUJKS128F4262F5F  ARJA1841187FB3A029       -1          5   
13     TRLLCAL128F428B903  ARP2RHS1187B991595       -1          5   
14     TREVWUZ128F4263A9B  AR9UYPT1187B9AE833       -1          6   
15     TRGYREY128E0791913  ARQ294N1187FB53D2A       -1          6   
16     TRBLBSR128F425EBFE  AR36FFP1187B9926D7       -1          7   
17     TRLRWJK128F427D602  AR6NBDC1187FB4D96D       -1          7   
18     TRYEKWE128F1459C42  AR6AK6B1187FB48EDD       -1          8   
19     TRPTALT128F426F99F  ARQRDXP1187B98B9E6       -1          8   
20     TRYXXWR12903CA7191  ARMBTFC1187FB56343       -1          9   
21     TRTTEXX128F428BE1D  ARJVH1S1187FB378C3       -1          9   
22     TRQPBQV128F423C1E9  AREH7LP1187B9B7164       -1          9   
23     TRVXQXV128F427863A  ARTO3L81187B9ACA8C       -1         10   
24     TRZJWSX12903CA89C1  ARC1IHZ1187FB4E920       -1         10   
25     TREVQFK128F4267577  ARTQ8QP1187FB3D220       -1         11   
26     TRFWVIP128EF3444F3  ARE66XI1187B9B8743       -1         11   
27     TRXZEBD128F4268BDB  AR33KSV1187FB4B900       -1         12   
28     TRBZYQA128F9359288  AR20SBM1187B9966A7       -1         12   
29     TRTLSVW12903CFB4F5  ARWX5TK1187FB4CDBE       -1         13   
...                   ...                 ...      ...        ...   
18166  TRDTJZI128EF34E78F  AR1BI4V1187B9937A9       -1       5842   
18167  TRJUXRE128F92F88D4  ARHSO041187FB3D06D    33329       5842   
18168  TRKFNUK128F428A62D  ARV3CRH1187B9A1B21    14338       5842   
18169  TROWHKP12903CCC703  ARCGI111187B9B8B99    67211       5843   
18170  TRELXLZ12903CB9DC5  ARYDLXP122988F049E    62368       5843   
18171  TRPMGHH128E07917AC  ARF4L041187FB4D318    99651       5844   
18172  TRMTFOJ12903CAFF57  AR3KVH41187FB3F71F    99652       5844   
18173  TRTWMBY128F9307A6D  ARSX6IX1187B9B640C    99655       5845   
18174  TRNNQSH128F4245865  AR4SL791187B9890BF       -1       5845   
18175  TRUAXLK12903CD39B5  AREYJKQ1187B9AEAB5   104087       5845   
18176  TRITRRU128F423DC76  ARJ59K61187FB5B43E    99662       5846   
18177  TRRCCUQ128F42449F7  ARNUU2R1187B9A316C    99675       5846   
18178  TRCEMLE128F4281036  ARJ59K61187FB5B43E    99668       5847   
18179  TRCGGYF128F9338B93  ARYDZKJ1187FB540C6    99680       5847   
18180  TRAHFYD128F426BBAD  AR94ZOI1187FB46BDA   103980       5848   
18181  TRRUFPY128F4261744  ARJWDFW1241B9C69AE   118487       5848   
18182  TRTFWRF128F92E378E  ARX1CYE1187FB3A97B    99730       5849   
18183  TRBXSHQ128F4278320  ARLM6I41187B9BA123    99731       5849   
18184  TRGOMBU128F42895B9  AR14CJ91187FB3A994    99733       5849   
18185  TRMCRMX128F4267AC3  ARSPUHE1187B99DB8A   125607       5850   
18186  TRNBSWN128F1473B31  ARLG9UJ1187B9B6C67    24567       5850   
18187  TRWAULS128F1474CBF  AR7GM0K1187B9B74B4    20304       5851   
18188  TRZVRJL128F92E2984  ARDYR3C1187FB461CE     9982       5851   
18189  TRMQPSQ128F426C548  ARAWPR61187B9B5A4C       -1       5851   
18190  TRSDUWY128F92DD39A  ARR176H1187FB4CDFE    99838       5852   
18191  TRUZAVE128F426E391  AR3DLBB1187B98F3DF   106651       5852   
18192  TRDNIRY12

Once all the files are accessible on the cluster, we will have to go through our SHS dataset and get those attributes for each track_id.
We will do so in the following way : (the paths are just examples on the subset)

In [72]:
tempo = []
hotness = []

for track in covers["trackID"]:
    folder1 = track[2]
    folder2 = track[3]
    folder3 = track[4]
    folder_path = "data/MillionSongSubset/data/" + folder1 + "/" + folder2 + "/" + folder3 + "/"
    track_path = folder_path + track + ".h5"
    if Path(track_path).exists(): #to delete later
        tempo.append(pd.read_hdf(track_path,"/analysis/songs")["tempo"][0])
        hotness.append(pd.read_hdf(track_path,"/metadata/songs")["song_hotttnesss"][0])

In [73]:
print(len(tempo))
print(np.sum(~np.isnan(hotness)))

204
128


Unsurprisingly, we only found 204 of those tracks in the subset and 128 of them have a hotness value.

### MILESTONE 3

In [46]:
dataset_path = "../../ruenz/millionsong_total.csv"
full_df = pd.read_csv(dataset_path,usecols=["track_id","tempo","song_hotttnesss"])

In [47]:
full_df

track_id  song_hotttnesss    tempo
0       TRAAARJ128F9320760   0.265861049211   86.643
1       TRAAAAK128F9318786   0.733371619962   99.944
2       TRAAAQO12903CD8E1C   0.254279270452  110.325
3       TRAAAAV128F421A322   0.481694146614  125.475
4       TRAAAUC128F428716F              nan  123.756
5       TRAAAAW128F429D538   0.602119989906   92.198
6       TRAAAUR128F428B1FA   0.767828738594  148.957
7       TRAAAAY128F42A73F0              0.0   41.279
8       TRAAAYL128F4271A5B   0.614752357554  127.503
9       TRAAABD128F429CF47              nan  121.274
10      TRAAAYX128F4263BC0              nan  179.677
11      TRAAACN128F9355673              nan  161.990
12      TRAAAZF12903CCCF6B              nan  110.129
13      TRAAACV128F423E09E              nan  100.812
14      TRAAAZU128F4226F7A   0.349119964652  202.363
15      TRAAADJ128F4287B47     0.3759843015  136.768
16      TRAAADT12903CCC339              0.0   97.623
17      TRAAADZ128F9348C2E              nan  100.070
18      TRAAAEA128F935A30D              0.0  156.677
19      TRAAAED128E0783FAB              nan  152.156
20      TRAAAEF128F4273421              nan  119.293
21      TRAAAEM128F93347B9   0.468997787645   91.947
22      TRAAAEW128F42930C0   0.407901717218  106.680
23      TRAAAFD128F92F423A   0.604500738589  129.738
24      TRAAAFI12903CE4F0E              nan  116.770
25      TRAAAFP128F931B4E3              nan  141.968
26      TRAAAFW128F42A4CFD              nan  119.027
27      TRAAAGF12903CEC202              nan   96.282
28      TRAAAGR128F425B14B              nan  123.989
29      TRAAAGW12903CC1049              nan  150.491
...                    ...              ...      ...
462516  TRLZZJW128F423ED3F   0.573626006665  112.150
462517  TRLZZKD12903CAF16F              nan  180.282
462518  TRLZZKV128F9317317              nan  110.005
462519  TRLZZLU128F4251B78              0.0  118.680
462520  TRLZZLX128F14591CC   0.396905220066  137.633
462521  TRLZZNA12903CCA082              nan  125.976
462522  TRLZZOA128F42322C7              nan  125.809
462523  TRLZZOJ128F1494C12              nan  134.807
462524  TRLZZOU128E078B2B1              nan   90.583
462525  TRLZZOZ12903CB30B8              nan  130.837
462526  TRLZZPI128F9314C80              nan   45.618
462527  TRLZZQV128E079896B   0.685274901698  101.053
462528  TRLZZRB128F932C5CB              nan   85.460
462529  TRLZZRN12903C9F7A3              nan  158.064
462530  TRLZZRS128EF34DAA4              nan  124.226
462531  TRLZZRW128F92F8B5F              nan  136.976
462532  TRLZZSG128F9330DD9              nan  121.927
462533  TRLZZSP128F93158B2    0.52976916466  160.096
462534  TRLZZTK128F425BC4C              nan  147.763
462535  TRLZZTN128F9343D1B    0.41967427881  170.046
462536  TRLZZUH128F9344865   0.249065794854  151.785
462537  TRLZZUR128F9308C99              nan  104.380
462538  TRLZZVP128F428551E              nan  119.920
462539  TRLZZVR128F931E5F2              nan  163.761
462540  TRLZZVS128F428CA4D              nan   97.615
462541  TRLZZWG128F425ABC5              nan   90.168
462542  TRLZZWQ128F9332435              nan  141.975
462543  TRLZZWU12903D0403B              nan  138.455
462544  TRLZZZF128F4278A7F              nan  205.083
462545  TRLZZZW12903CC8212              nan  140.038

[462546 rows x 3 columns]

In [48]:
full_df[full_df.song_hotttnesss == "nan"]="NaN"

In [61]:
full_df.count()

track_id           462546
song_hotttnesss    462546
tempo              462546
dtype: int64

In [50]:
#percentage of song_hotttnesss available
100 * ( 1 - full_df.song_hotttnesss.str.contains("nan").sum() / full_df.shape[0] )

57.900835808762793

In [59]:
#percentage of tempo available
100 * ( 1 - (full_df.tempo==0).sum() / full_df.shape[0] )

99.662087662632473

In [56]:
#Number of matches between the full set and the cover set
len(set(full_df.track_id) & set(covers.trackID))

8689

### SUITE DU PIPELINE

* détermination de l'original en fonction de l'année pour les unknown :

In [178]:
covers = pickle.load(open("covers_merged.p","rb"))
covers

trackID            artistID  shsPerf  clique_id  \
0      TRGDMZP128F42BC52B  ARB1DDF1187FB4FCFB        0          0   
1      TRCATYW12903D038FE  ARGJEEO1271F573FD6        0          0   
2      TRVMZJZ128F4270CE4  ARY0HTV1187FB4A1B1   412972          0   
3      TRKOINL128F42926C3  ARQ5FSZ1187B98AD74        0          0   
4      TROJZTF128F428B546  ARJN76O1187FB43C99        0          1   
5      TRYQEDQ128F427917C  ARS4KT21187B9B9438        0          1   
6      TRCKNGE128F92DA3F3  AR1CB5G1187B9AFB8E    16660          2   
7      TRIOPLY128F423CFF3  ARKZJ301187FB521B2   551633          2   
8      TRWNDEU128F9329BF7  ARVZWQ31187B9B8946   354066          3   
9      TRYOPHS128F146DEFD  AR6NYHH1187B9BA128   354067          3   
10     TRMBSQR128F92DF66E  ARPI2DX1187FB4CED4    52010          4   
11     TRQNZCE128E078A9C0  ARWILYB1187FB37DFE    52011          4   
12     TRTUJKS128F4262F5F  ARJA1841187FB3A029        0          5   
13     TRLLCAL128F428B903  ARP2RHS1187B991595        0          5   
14     TREVWUZ128F4263A9B  AR9UYPT1187B9AE833    45769          6   
15     TRGYREY128E0791913  ARQ294N1187FB53D2A     9133          6   
16     TRBLBSR128F425EBFE  AR36FFP1187B9926D7    59663          7   
17     TRLRWJK128F427D602  AR6NBDC1187FB4D96D      575          7   
18     TRYEKWE128F1459C42  AR6AK6B1187FB48EDD    68273          8   
19     TRPTALT128F426F99F  ARQRDXP1187B98B9E6   305654          8   
20     TRYXXWR12903CA7191  ARMBTFC1187FB56343   160334          9   
21     TRTTEXX128F428BE1D  ARJVH1S1187FB378C3        0          9   
22     TRQPBQV128F423C1E9  AREH7LP1187B9B7164    14943          9   
23     TRVXQXV128F427863A  ARTO3L81187B9ACA8C    21513         10   
24     TRZJWSX12903CA89C1  ARC1IHZ1187FB4E920    21514         10   
25     TREVQFK128F4267577  ARTQ8QP1187FB3D220    31718         11   
26     TRFWVIP128EF3444F3  ARE66XI1187B9B8743    31719         11   
27     TRXZEBD128F4268BDB  AR33KSV1187FB4B900    88882         12   
28     TRBZYQA128F9359288  AR20SBM1187B9966A7    88880         12   
29     TRTLSVW12903CFB4F5  ARWX5TK1187FB4CDBE    29876         13   
...                   ...                 ...      ...        ...   
18166  TRDTJZI128EF34E78F  AR1BI4V1187B9937A9   184203       5842   
18167  TRJUXRE128F92F88D4  ARHSO041187FB3D06D    33329       5842   
18168  TRKFNUK128F428A62D  ARV3CRH1187B9A1B21    14338       5842   
18169  TROWHKP12903CCC703  ARCGI111187B9B8B99    67211       5843   
18170  TRELXLZ12903CB9DC5  ARYDLXP122988F049E    62368       5843   
18171  TRPMGHH128E07917AC  ARF4L041187FB4D318    99651       5844   
18172  TRMTFOJ12903CAFF57  AR3KVH41187FB3F71F    99652       5844   
18173  TRTWMBY128F9307A6D  ARSX6IX1187B9B640C    99655       5845   
18174  TRNNQSH128F4245865  AR4SL791187B9890BF        0       5845   
18175  TRUAXLK12903CD39B5  AREYJKQ1187B9AEAB5   104087       5845   
18176  TRITRRU128F423DC76  ARJ59K61187FB5B43E    99662       5846   
18177  TRRCCUQ128F42449F7  ARNUU2R1187B9A316C    99675       5846   
18178  TRCEMLE128F4281036  ARJ59K61187FB5B43E    99668       5847   
18179  TRCGGYF128F9338B93  ARYDZKJ1187FB540C6    99680       5847   
18180  TRAHFYD128F426BBAD  AR94ZOI1187FB46BDA   103980       5848   
18181  TRRUFPY128F4261744  ARJWDFW1241B9C69AE   118487       5848   
18182  TRTFWRF128F92E378E  ARX1CYE1187FB3A97B    99730       5849   
18183  TRBXSHQ128F4278320  ARLM6I41187B9BA123    99731       5849   
18184  TRGOMBU128F42895B9  AR14CJ91187FB3A994    99733       5849   
18185  TRMCRMX128F4267AC3  ARSPUHE1187B99DB8A   125607       5850   
18186  TRNBSWN128F1473B31  ARLG9UJ1187B9B6C67    24567       5850   
18187  TRWAULS128F1474CBF  AR7GM0K1187B9B74B4    20304       5851   
18188  TRZVRJL128F92E2984  ARDYR3C1187FB461CE     9982       5851   
18189  TRMQPSQ128F426C548  ARAWPR61187B9B5A4C     9981       5851   
18190  TRSDUWY128F92DD39A  ARR176H1187FB4CDFE    99838       5852   
18191  TRUZAVE128F426E391  AR3DLBB1187B98F3DF   106651       5852   
18192  TRDNIRY12

In [179]:
# general stat
covers[covers["original_id"] == "Unknown"]

trackID            artistID  shsPerf  clique_id  \
0      TRGDMZP128F42BC52B  ARB1DDF1187FB4FCFB        0          0   
1      TRCATYW12903D038FE  ARGJEEO1271F573FD6        0          0   
2      TRVMZJZ128F4270CE4  ARY0HTV1187FB4A1B1   412972          0   
3      TRKOINL128F42926C3  ARQ5FSZ1187B98AD74        0          0   
4      TROJZTF128F428B546  ARJN76O1187FB43C99        0          1   
5      TRYQEDQ128F427917C  ARS4KT21187B9B9438        0          1   
12     TRTUJKS128F4262F5F  ARJA1841187FB3A029        0          5   
13     TRLLCAL128F428B903  ARP2RHS1187B991595        0          5   
35     TRUATRC12903CAAF21  AR0RCMP1187FB3F427        0         16   
36     TRPJVKE128F428AA27  ARCCN681187B9AFD73        0         16   
45     TRXMDOP12903CBC46A  ARKG7TE1187B9A45D8        0         21   
46     TRKEXQZ128F14959B0  ARJFOC01187FB413A5    60152         21   
47     TRRYIVT128F9336500  AR5HOKG1187FB3873E    60137         21   
52     TRDIRXL128F4222742  ARG81VK1187FB401BF        0         24   
53     TRJZZNO128F92F48DE  ARXFMP01187FB40768        0         24   
54     TRBESBZ12903CA0A95  ARR2BHS1187B9B8FA7        0         25   
55     TROFEEX128F92F190C  AR5JUO91187FB4E90B        0         25   
74     TRSWHFV128F421A882  ARZ7MPB1187B98CA77        0         34   
75     TRHXZRK12903D02C2D  ARH861H1187B9B799E        0         34   
76     TRPKERG128F92CA8F1  AR5US1D1187B9AEB1F        0         35   
77     TRRFVJY128EF3523EC  AR8OVZ01187B9A5E1A        0         35   
86     TRKFOUP128E0786014  ARUJZFJ1187B9B135F        0         40   
87     TRRLKSI128F92D5D37  ARWQ4AM1187B9AD5F8        0         40   
88     TRAOGDO128F42B9A6F  ARPYV1F1187FB444FA        0         41   
89     TRNUMMX128F92D03F4  ARZHQDD1187FB5B871        0         41   
95     TRCLFJM128F92FE654  ARRS1FT1187B9B650A        0         44   
96     TRECEYT12903CB8F86  AR7KA5V1187FB44E6B        0         44   
99     TRYVXMT128F4272422  AREQIT21187B9AEB7F        0         46   
100    TREEKRX128E0790AD3  AR6TNOA1187B9B5C44        0         46   
103    TRGEHDP128F148A1B4  AR9YWMS1187FB43A34   319149         48   
...                   ...                 ...      ...        ...   
18056  TRFZDMW12903CAB4AE  AR772T41187B98E14F   111345       5800   
18057  TRQENZL128F92EBCC9  AR9F71I1187FB46CAF    98385       5800   
18058  TRELMUN128F4221976  AR5XYFL1187FB414C0    98322       5801   
18059  TRHWDRI128F92CE267  AR772T41187B98E14F   146420       5801   
18060  TRXJVYT128F42345CA  ARD0FN11187B9B25CD   101437       5801   
18067  TRRFEGU128F9335471  AREUZEN1187FB48155    85738       5805   
18068  TRJIAUW128F4253585  ARI5E6U1187B9B525A    45026       5805   
18089  TRORIEA128F42B155F  AR4ZNBW1187FB554B8   148752       5814   
18090  TRBZFDK128EF35A246  ARN0YZ71187B9BA57B   109617       5814   
18095  TRCCGSP12903CDC426  AR9VN011187B9ADD25    86489       5817   
18096  TRNSTHZ128F149EBAC  ARB336A1187B9A4583     9880       5817   
18104  TRHBRQQ12903CCA3D6  ARDHAOT11F4C83C0BC    25364       5821   
18105  TRDZUEO128F92E40D1  AREI6PI1187B9911AE    15971       5821   
18106  TRBIKMD128F145C451  ARGGLM71187B9AA273    16179       5822   
18107  TREFDKH128F93135E9  ARBOG0H1187B99F775    91356       5822   
18134  TRYTOUE128F426599B  ARO0QD41187FB3AB0A    15798       5830   
18135  TRQUJCP128F4238206  ARLAUED1187B9ACEAF    16397       5830   
18136  TRPMDDK12903D0D320  AREWAAS1187B98C490        0       5830   
18169  TROWHKP12903CCC703  ARCGI111187B9B8B99    67211       5843   
18170  TRELXLZ12903CB9DC5  ARYDLXP122988F049E    62368       5843   
18180  TRAHFYD128F426BBAD  AR94ZOI1187FB46BDA   103980       5848   
18181  TRRUFPY128F4261744  ARJWDFW1241B9C69AE   118487       5848   
18185  TRMCRMX128F4267AC3  ARSPUHE1187B99DB8A   125607       5850   
18186  TRNBSWN128F1473B31  ARLG9UJ1187B9B6C67    24567       5850   
18190  TRSDUWY128F92DD39A  ARR176H1187FB4CDFE    99838       5852   
18191  TRUZAVE128F426E391  AR3DLBB1187B98F3DF   106651       5852   
18192  TRDNIRY12

In [127]:
# The years recovered from the dataset are floats, so we need to strip their decimal part,
# and then we can take the last 4 characters on any date string to get the year.
covers["year"] = covers["date"].replace("\.0","",regex=True).str[-4:]

In [132]:
# In each clique, we rank the versions by year to find the original
covers["rank"] = covers.groupby("clique_id")["year"].rank(method="dense",ascending=False).astype(int)

In [128]:
covers

trackID            artistID  shsPerf  clique_id  \
0      TRGDMZP128F42BC52B  ARB1DDF1187FB4FCFB        0          0   
1      TRCATYW12903D038FE  ARGJEEO1271F573FD6        0          0   
2      TRVMZJZ128F4270CE4  ARY0HTV1187FB4A1B1   412972          0   
3      TRKOINL128F42926C3  ARQ5FSZ1187B98AD74        0          0   
4      TROJZTF128F428B546  ARJN76O1187FB43C99        0          1   
5      TRYQEDQ128F427917C  ARS4KT21187B9B9438        0          1   
6      TRCKNGE128F92DA3F3  AR1CB5G1187B9AFB8E    16660          2   
7      TRIOPLY128F423CFF3  ARKZJ301187FB521B2   551633          2   
8      TRWNDEU128F9329BF7  ARVZWQ31187B9B8946   354066          3   
9      TRYOPHS128F146DEFD  AR6NYHH1187B9BA128   354067          3   
10     TRMBSQR128F92DF66E  ARPI2DX1187FB4CED4    52010          4   
11     TRQNZCE128E078A9C0  ARWILYB1187FB37DFE    52011          4   
12     TRTUJKS128F4262F5F  ARJA1841187FB3A029        0          5   
13     TRLLCAL128F428B903  ARP2RHS1187B991595    73986          5   
14     TREVWUZ128F4263A9B  AR9UYPT1187B9AE833    45769          6   
15     TRGYREY128E0791913  ARQ294N1187FB53D2A        0          6   
16     TRBLBSR128F425EBFE  AR36FFP1187B9926D7    59663          7   
17     TRLRWJK128F427D602  AR6NBDC1187FB4D96D      575          7   
18     TRYEKWE128F1459C42  AR6AK6B1187FB48EDD    68276          8   
19     TRPTALT128F426F99F  ARQRDXP1187B98B9E6   305654          8   
20     TRYXXWR12903CA7191  ARMBTFC1187FB56343   160334          9   
21     TRTTEXX128F428BE1D  ARJVH1S1187FB378C3        0          9   
22     TRQPBQV128F423C1E9  AREH7LP1187B9B7164    14943          9   
23     TRVXQXV128F427863A  ARTO3L81187B9ACA8C    21513         10   
24     TRZJWSX12903CA89C1  ARC1IHZ1187FB4E920    21514         10   
25     TREVQFK128F4267577  ARTQ8QP1187FB3D220    31718         11   
26     TRFWVIP128EF3444F3  ARE66XI1187B9B8743    31719         11   
27     TRXZEBD128F4268BDB  AR33KSV1187FB4B900    88882         12   
28     TRBZYQA128F9359288  AR20SBM1187B9966A7    88880         12   
29     TRTLSVW12903CFB4F5  ARWX5TK1187FB4CDBE    29876         13   
...                   ...                 ...      ...        ...   
18166  TRDTJZI128EF34E78F  AR1BI4V1187B9937A9   184203       5842   
18167  TRJUXRE128F92F88D4  ARHSO041187FB3D06D    33329       5842   
18168  TRKFNUK128F428A62D  ARV3CRH1187B9A1B21    14338       5842   
18169  TROWHKP12903CCC703  ARCGI111187B9B8B99    67211       5843   
18170  TRELXLZ12903CB9DC5  ARYDLXP122988F049E    62368       5843   
18171  TRPMGHH128E07917AC  ARF4L041187FB4D318    99651       5844   
18172  TRMTFOJ12903CAFF57  AR3KVH41187FB3F71F    99652       5844   
18173  TRTWMBY128F9307A6D  ARSX6IX1187B9B640C    99655       5845   
18174  TRNNQSH128F4245865  AR4SL791187B9890BF        0       5845   
18175  TRUAXLK12903CD39B5  AREYJKQ1187B9AEAB5   104087       5845   
18176  TRITRRU128F423DC76  ARJ59K61187FB5B43E    99662       5846   
18177  TRRCCUQ128F42449F7  ARNUU2R1187B9A316C    99675       5846   
18178  TRCEMLE128F4281036  ARJ59K61187FB5B43E    99668       5847   
18179  TRCGGYF128F9338B93  ARYDZKJ1187FB540C6    99680       5847   
18180  TRAHFYD128F426BBAD  AR94ZOI1187FB46BDA   103980       5848   
18181  TRRUFPY128F4261744  ARJWDFW1241B9C69AE   118487       5848   
18182  TRTFWRF128F92E378E  ARX1CYE1187FB3A97B    99730       5849   
18183  TRBXSHQ128F4278320  ARLM6I41187B9BA123    99731       5849   
18184  TRGOMBU128F42895B9  AR14CJ91187FB3A994    99733       5849   
18185  TRMCRMX128F4267AC3  ARSPUHE1187B99DB8A   125607       5850   
18186  TRNBSWN128F1473B31  ARLG9UJ1187B9B6C67    24567       5850   
18187  TRWAULS128F1474CBF  AR7GM0K1187B9B74B4    20304       5851   
18188  TRZVRJL128F92E2984  ARDYR3C1187FB461CE     9982       5851   
18189  TRMQPSQ128F426C548  ARAWPR61187B9B5A4C     9981       5851   
18190  TRSDUWY128F92DD39A  ARR176H1187FB4CDFE    99838       5852   
18191  TRUZAVE128F426E391  AR3DLBB1187B98F3DF   106651       5852   
18192  TRDNIRY12

In [136]:
#number of first version
(covers["rank"] == 1).sum()

6209

In [141]:
#number of cliques
covers.clique_id.max() + 1

5854

In [149]:
unknown_originals = covers[covers["rank"] == 1][covers[covers["rank"] == 1].duplicated(subset="clique_id",keep=False)]

In [171]:
covers[covers["year"].str.contains("nan") & covers["original_id"].str.contains("Unknown")].shape

(154, 13)

In [173]:
covers.drop(covers[covers["year"].str.contains("nan") & covers["original_id"].str.contains("Unknown")].index).shape

(18042, 13)

In [180]:
len(covers[covers.duplicated('shsPerf',keep=False) & covers.shsPerf!=0].sort_values('shsPerf'))

261